### ISSUE TITLE
arrayBuffer fetch type not documented in the fetch chunk section of the docs

### ISSUE THREAD

**Issue by hamilton** [iodide-project/iodide#1686]
~Array Buffers are pretty common response formats for `fetch` calls in JS. At the moment the closest we have is `blob` which requires some gymnastics to get into the right format. Critically, I think this would prevent us from exploiting some of the more interesting binary-based data representations we care about, including apache arrow tables.~

~This would make a really great first issue, since it just requires that we use `.arrayBuffer()`  on the `Response` object generated in the fetch actions and add the logic and syntax to handle that.~

(NB: this issue was closed with a recent PR. We need to add `arrayBuffer` to the docs) and then we can close this one. Updating the title to reflect this)

**Comment by hamilton** [comment]
This was actually closed in a recent PR (that was much bigger than this small issue) but our docs have not been updated accordingly. I'm going to update this to say we need to add to the docs.

**Comment by vadimkantorov** [comment]
I've just tried out `arrayBuffer` mode on `definitely-staging.iodide.io`: the `arrayBuffer` keyword is not hightlighted as is `blob`

**Comment by vadimkantorov** [comment]
Also there seem some correctness issues: https://definitely-staging.iodide.io/notebooks/37/ (the new way returns some bad bytes, the old blob+FileReader way works fine)

**Comment by hamilton** [comment]
Yes, syntax highlighting hasn't been implemented quite yet on `arrayBuffer` - thanks for catching that. Should be easy to add.

Re: your example, you would still need to wrap the output of the `arrayBuffer` in `Uint8Array`, right? That particular conversion is not automatic w/ the fetch chunk type. All `arrayBuffer` does is effectively gives you `fetch(url).then(r=>r.arrayBuffer())` so you would probably need to do additional conversion if that's necessary. That could change down the line if needed but I think we'd probably prefer to keep it as plain as possible so that we can be as flexible as we need here.

**Comment by vadimkantorov** [comment]
at least on Python's side both seem to have the same type:
```
type(electionsData_memoryview2)
<class 'memoryview'>
type(electionsData_memoryview)
<class 'memoryview'>
```

**Comment by vadimkantorov** [comment]
And even if this is right, the task of binary dataset loading is so common (given super-slowness of pandas and csv parsing), that there should be a way without three different magics: `fetch`, (tricky) `js` and `py`

**Comment by hamilton** [comment]
It's not really about type. Your two examples aren't the same because in one you're doing an additional coercion to `Uint8Array`. `arrayBuffer` returns an array buffer, _not_ a typed array. I performed the additional `Uint8Array` call as you did in the `blob` example and it seems to work as expected (no error, that is).

**Comment by hamilton** [comment]
I agree this can be potentially confusing to users, though, so we'll have to (1) provide some example notebooks to make it clear how to do these sorts of things and (2) think a bit more about how to smooth out the process.

**Comment by vadimkantorov** [comment]
Maybe an extra fetch type `bytes` or `binary` (that would be convertable to uint8)?

**Comment by hamilton** [comment]
Yeah, that's not a bad idea, and quite simple to implement. To be frank the majority of time I've used array buffers has been loading data that requires `Uint8Array` (arrow, etc.) so I feel your pain. I'll file a separate issue for that one and chat w/ the team about it.

**Comment by vadimkantorov** [comment]
Meanwhile, should `js.Uint8Array.new(js.myarraybuffer)` theoretically work? (once `import js` + `__getattr__` PR lands)

**Comment by hamilton** [comment]
In theory that should work!

**Comment by vadimkantorov** [comment]
How did you do the trick with `new Uint8Array(...)`?

For me the conversion hangs forever:
```
%% fetch
arrayBuffer: electionsData_memoryview2=https://cors-anywhere.herokuapp.com/https://github.com/schitaytesami/lab/releases/download/data/electionsData.npz
%% js
electionsData_memoryview3 = new Uint8Array(electionsData_memoryview2); # hangs forever
```

**Comment by hamilton** [comment]
I did the following and didn't experience any hang (though I haven't verified the data at all):

```
%% fetch
arrayBuffer: electionsData_memoryview2=https://cors-anywhere.herokuapp.com/https://github.com/schitaytesami/lab/releases/download/data/electionsData.npz

%% js
var electionsData = new Uint8Array(electionsData_memoryview2)

%% py
import numpy as np
import io
from js import electionsData

np.load(io.BytesIO(electionsData)) # completes OK

```

**Comment by vadimkantorov** [comment]
The difference is that you have `var`. Without `var` it hangs forever (at least in Chrome)

**Comment by hamilton** [comment]
Yes, just verified that in FF. That is most certainly a javascript language quirk! Yet another reason why having a `binary` fetch type would make things easier.

**Comment by vadimkantorov** [comment]
If it's consistent across browsers it'a a nastier quirk! Probably worth a documentation note somewhere :) or a deeper investigation (does it repro outside Iodide context with pure JavaScript?)

**Comment by hamilton** [comment]
I can say pretty confidently that it's a language quirk, and all JS engines implement it because whatever it's doing, it's baked into the agreed-upon ECMAScript spec. We are basically using the vanilla `eval` in JS, which means that whatever engine your browser comes with, it uses to run the javascript.

Javascript is a strange language in a lot of ways. Most js linters would throw an error if you wrote `data = ...` and it is heavily frowned-upon generally to not use `var`, `let`, or `const`. In our case, `var` is useful because of how it scopes outside of the chunk it is running in.

Yet another reason to make sure we have more affordances for Python natives in Iodide to do what they want to do without ever touching JS!

**Comment by vadimkantorov** [comment]
Also a question: is it meaningful to convert `arrayBuffer` as Python's `memoryview` (without a specified typed array conversion)?  What's the semantics?

**Comment by vadimkantorov** [comment]
I agree about having `var`'s are good (but it wasn't intuitive what is scoping of code cells, should I create "global" variables or if `var` is fine)!

Yet ```a = new Uint8Array([1,2,3])``` works without hang, so there is something about the interplay with the `arrayBuffer`.

**Comment by mdboom** [comment]
> Also a question: is it meaningful to convert arrayBuffer as Python's memoryview (without a specified typed array conversion)? What's the semantics?

An array buffer is a buffer of bytes, with no real semantics beyond this.  So a Python `memoryview` is a pretty close analogy.  To interpret it as anything else on the Python side, you'd cast to `bytes` or a numpy array with a specific dtype or something...

**Comment by vadimkantorov** [comment]
Confused by the `memoryview` type I tried to pipe it in BytesIO (as you've seen), but it seemed that it's not a simple array of bytes, otherwise it'd have worked and Uint8Array conversion would not have been needed.

Also the following fails (with the same old numpy.load error):
```
%% fetch
arrayBuffer: data=https://cors-anywhere.herokuapp.com/https://github.com/schitaytesami/lab/releases/download/data/electionsData.npz

%% py
import numpy as np
import io
from js import data, Uint8Array
np.load(io.BytesIO(Uint8Array.new(data))) # error
```

**Comment by vadimkantorov** [comment]
@hamilton Should I create a separate issue for the last snippet?

**Comment by hamilton** [comment]
@vadimkantorov that sounds good. It sounds like a pyodide-specific issue so I would file it in that repository specifically.

**Comment by vadimkantorov** [comment]
thanks for assistiance!

( created https://github.com/iodide-project/pyodide/issues/406 )

**Comment by mdboom** [comment]
I think the hanging comes not from a language quirk but from trying to display a really large typed array to the console.  I filed #1787 for that.

**Comment by hamilton** [comment]
this is so obvious in hindsight! Thanks @mdboom.
